velox/vector/tests/VectorRetainedSizeTest.cpp

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include "velox/vector/tests/utils/VectorTestBase.h"
+
+using namespace facebook;
+using namespace facebook::velox;
+
+class VectorRetainedSizeTest : public testing::Test,
+                               public velox::test::VectorTestBase {
+ protected:
+  static void SetUpTestCase() {
+    memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  }
+
+  std::vector<std::string> longStrings_{
+      {std::string(100, 'a'), std::string(200, 'b'), std::string(300, 'c')}};
+};
+
+namespace {
+StringView shortStringAt(vector_size_t row) {
+  return StringView::makeInline(std::string(row % 3, '.'));
+}
+} // namespace
+
+TEST_F(VectorRetainedSizeTest, flatVectorNoStrings) {
+  auto vector = makeFlatVector<int32_t>(1'000, [](auto row) { return row; });
+  uint64_t totalStringBufferSize = 0;
+  auto retainedSize = vector->retainedSize(totalStringBufferSize);
+
+  EXPECT_EQ(totalStringBufferSize, 0);
+  EXPECT_EQ(retainedSize, vector->retainedSize());
+}
+
+TEST_F(VectorRetainedSizeTest, flatVectorInlinedStrings) {
+  auto vector = makeFlatVector<StringView>(1'000, shortStringAt);
+  uint64_t totalStringBufferSize = 0;
+  auto retainedSize = vector->retainedSize(totalStringBufferSize);
+
+  EXPECT_EQ(totalStringBufferSize, 0);
+  EXPECT_EQ(retainedSize, vector->retainedSize());
+}
+
+TEST_F(VectorRetainedSizeTest, flatVectorMultipleStringBuffers) {
+  auto vector = makeFlatVector<StringView>({});
+
+  std::string str1(100, 'a');
+  std::string str2(200, 'b');
+  std::string str3(150, 'c');
+
+  vector->resize(3);
+  auto* buf1 = vector->getBufferWithSpace(str1.size());
+  memcpy(buf1->asMutable<char>(), str1.data(), str1.size());
+  vector->setNoCopy(0, StringView(buf1->as<char>(), str1.size()));
```

Comment:
why not just vector->set(0, value)?

Reply:
@mbasmanova  Because I would like to make the vector has multiple string buffers. With vector->set(0, value), it's possible that all three strings are put in the same string buffer.